### ISSUE TITLE
Support for Open CL

### ISSUE THREAD

**Issue by agibsonccc** [deeplearning4j/nd4j#27]
This will be the thread for AMD gpus.

**Comment by agibsonccc** [comment]
Initial WIP happened. Converted most .cu files to .kh and mapped the terminology. Next will be integrating jocl.

**Comment by peterstyles** [comment]
+1 for this - cheers for putting this all together!

**Comment by agibsonccc** [comment]
Quick update on this: cuda architecture is baked.

I am going to base the opencl support off of the stuff I did in cuda. The
opencl kernels are written already. Just need to finish porting the cuda
stuff over now.

**Comment by Riotsup** [comment]
Hello is this already supported for AMD?
Cause, im planning on buying a Fury X for dl and for the lulz.

**Comment by agibsonccc** [comment]
Opencl supports AMD yes. I need to port the work I did for cuda to opencl though. The kernels are already written. Need bandwidth though.

**Comment by Riotsup** [comment]
Great, but what is it that nd4j does not do for AMD that it does for CUDA (yet)?

**Comment by agibsonccc** [comment]
Nd4j cuda is actually implemented. I don't have the Blas operations for
opencl yet.

**Comment by agibsonccc** [comment]
Of note: https://github.com/clMathLibraries/clSPARSE/

**Comment by rkraneis** [comment]
Hi, what is the current state of this? Manually compiling against clBLAS? nd4j-0.4 seems to have dropped anything Jocl ...

**Comment by agibsonccc** [comment]
We never really finished it.  We don't have customers asking for it. You
are welcome to send us a pull request (hint: no one steps up)

If you want it we will direct you. Beyond that we don't have bandwidth.

**Comment by raver119** [comment]
Just a side note: openCL is viable for Xeon Phi...

**Comment by agibsonccc** [comment]
Problem is versioning. What versions do we support? Where do we expect it to run? WHAT should we support? There's an insane amount of fragmentation out there.

**Comment by raver119** [comment]
Yea, but right now i'm not mentioning AMD gpus at all, AMD lost that battle and i'm fine with that. Was speaking only about Xeon Phi, and their use with nd4j :)

**Comment by saudet** [comment]
AFAIK, Intel is pushing more for OpenMP than OpenCL there, e.g.:

> Q3: how is working with MIC/GPU compared to OpenCL in terms of performance, and coding comlexity?
> A3: Good question, the feedback we got from customers wwere, e.g. Given one large C++ applications, it takes 6 monthes to re-implement it with OpenCL, but, it takes about 6 weeks using OpenMP with code modernization efforts to achieve a better perofrmance. In addition, based on a set of workload, OpenCL performance is ~10-30% below OpenMP code on MIC.

https://software.intel.com/en-us/videos/new-era-for-openmp-beyond-traditional-shared-memory-parallel-programming

**Comment by rkraneis** [comment]
I'm mostly interested in increased performance of deeplearning4j on Amd/Intel GPUs. Are there benchmarks on the improvements for e.g. intel hd over core i7  (skylake)? Or is it just not worth the hassle?

@saudet: so you would recommend implementing nd4j on OpenMP rather OpenCL at least for Intel GPUs?

**Comment by agibsonccc** [comment]
We already did openmp.

**Comment by rkraneis** [comment]
@agibsonccc, do you mean the "Linking with MKL" part?

**Comment by agibsonccc** [comment]
Right. So we already have mkl in place

**Comment by saudet** [comment]
@rkraneis Xeon Phi is more CPU than GPU, and in that case Intel seems to lean toward OpenMP yes. We'll need to optimize for the Xeon Phi obviously though. As for actual GPUs from Intel, those will require OpenCL, just like AMD. Although if OpenCL ends up being only useful in the case of GPUs, there's also Vulkan we should take a look at...

**Comment by raver119** [comment]
I'm afraid that "optimize for Xeon Phi" will end up with memory management, since auto offload mode will hit into PCIe transfers price. It's the same as for CUDA - nd4j issues many atomic operations, on relatively small data That's why i've mentioned OpenCL there...

**Comment by bhack** [comment]
They are pushing on openmp for [mkl-dnn](https://software.intel.com/en-us/articles/deep-neural-network-technical-preview-for-intel-math-kernel-library-intel-mkl) and They will opensource the dnn component in Q3.

**Comment by Lord-of-the-Galaxy** [comment]
Is this still on?

**Comment by AlexDBlack** [comment]
@Lord-of-the-Galaxy It's not on the immediate roadmap - it's a low priority for us currently (and would require a lot of engineering time), given that the main use case for OpenCL (i.e., AMD GPUs) have very little use in commercial contexts or cloud computing (AWS, Azure, etc).

**Comment by Lord-of-the-Galaxy** [comment]
I see. I thought OpenCL may gain priority as neural networks are starting to get used in more and more programs that run on the consumer's device (there are some games using neural networks extensively), but I guess I was wrong.

**Comment by agibsonccc** [comment]
NVIDIA effectively has a chokehold on the space. Not only are they giving away gpus, they are also  involved in integration in to a lot of the open source packages (us notwithstanding, we write our own).

Honestly until mesos recognizes what an AMD gpu is it's not reallly interesting for us.

Opencl is a horribly fragmented standard where it would mean different things on different devices.

NVIDIA still has their opencl support on 2.0. The only reason amd even pushes "open" is because hte are the underdog. A lot of HPC code also doesn't tend to need to be "cross platform".

"Cross platform" tends to be  "nvidia". As far as other kinds of devices are concerned like android it's a lot of the same story.

**Comment by raver119** [comment]
Right, we can add support for OpenCL, all our important stuff is c++ anyway, but i'm afraid  fragmentation will make it nightmare for users.

**Comment by xvolks** [comment]
The upcoming iMacPro may be worth the effort to use DL4j on AMD GPUs. But Apple may decide to implement a Keras/Caffe for this computer.

**Comment by agibsonccc** [comment]
@xvolks So 1 thing here: apple implemented a "backend" called coreml with support for various framework formats. CoreML is mainly for inference on iphones right now.

Do some research before you make claims please :).

**Comment by xvolks** [comment]
Hey, it was not a claim. I know CoreML is not for training. I just stated that the upcoming iMacPro should be a good candidate to make some training with DL4j if OpenCL backend was available. Sorry for the misunderstanding as English is not my primary language. 

Envoyé de mon iPhone

**Comment by InonS** [comment]
I'm not sure where things stand right now. `README.md` [module section](https://github.com/deeplearning4j/nd4j#modules) clearly states:

> Several of these modules are different backend options for ND4J (including GPUs).
> ...
> jocl-parent = Java bindings for OpenCL

from which it is understood that ND4J *currently* supports an OpenCL backend via JOCL.

However, looking into the [ND4J's backend implementations package](https://github.com/deeplearning4j/nd4j/tree/master/nd4j-backends/nd4j-backend-impls), I see only native or CUDA.

What happened to JOCL?

**Comment by agibsonccc** [comment]
Apologies, that's actually out of date. We had started an implementation of that a while back and ended up cancelling after seeing what it would take to build.

**Comment by Lord-of-the-Galaxy** [comment]
And any plans to attempt to use VULKAN (Is it even possible? I'm mostly uninformed on the topic)?

**Comment by agibsonccc** [comment]
We're really only looking at cuda for the short term.

**Comment by raver119** [comment]
at this moment of time vulkan isn't really suited for math. at least that's what I've got after reading it's documentation. it's tailored for graphics mostly, and future implementations might change that

**Comment by hristo-vrigazov** [comment]
I have seen people build pretty decent machines in terms of teraflops / dollar with AMD GPUs, but unforturnately, that forces you to use some Caffe fork since it is the only one that supports them. Check this post out: https://medium.com/intuitionmachine/building-a-50-teraflops-amd-vega-deep-learning-box-for-under-3k-ebdd60d4a93c

**Comment by treo** [comment]
Due to the current crypto mining craze, I would actually argue that - at least for single precision performance - AMD isn't that much better, and at 8GB, while HBM2, are pretty small.

The half precision story is a bit different, but I'd argue that it still isn't enough to make supporting it worth the effort.

**Comment by pshirshov** [comment]
Eh, guys, I assume there is no chance for OpenCL/ROCm support? :(

**Comment by aminnasiri** [comment]
Hey, Is there any chance to work on this ticket?

**Comment by saudet** [comment]
BTW, if you only need inference, the TVM backend works fine with OpenCL, ROCm, CoreML, Vulkan, etc.
